### DISCUSSION TITLE
useFormContext w/ reset functionality

### DISCUSSION THREAD

**Discussion by thooverdev** [react-hook-form/react-hook-form#8146] [Q&A] [answered]
Hello all,

I'm new to react-hook-form and trying to build reusable (DRY) fields that uses the FormProvider and useFormContext(). Registering works (assuming it does since when I click submit, the form data shows up in the HTMLTable) but for some reason methods.reset() isn't working and I can't figure out why...

Versions:
React: 17.0.2
@blueprintjs/core: 4.0.4
react-hook-form: 7.7.1

csb link: https://codesandbox.io/s/eloquent-platform-0l1n0m?file=/src/App.tsx

Any thoughts or suggestions would be welcome.

NOTE: You might look at this and think why is it so deeply nested. There are a lot more pieces than this at play so I tried simplifying it as best I could to show just the basic example of what's occurring and the setup I have.

NOTE #2: I don't know if this works in the sandbox environment, kind of a "fingers crossed it works for someone else" moment. My company computer blocks the embedded browser within csb but in theory, it should work.

**Comment by Moshyfawn** [comment, marked as answer]
You're spreading the `register` props into your `CustomTextField` component,
```js
<CustomTextField {...methods.register(identifier)} />
```
but trying to get the same `register` method inside the component instead of the props.
```js
const CustomTextField = (props: any) => {
  const { register } = props
}
```

So, basically, it should be something like
```diff
- const CustomTextField = (props: any) => {
+ const CustomTextField = (props: UseFormRegisterReturn) => {
-  const { register } = props
-  return <InputGroup {...register} />
+  return <InputGroup {...props} />
 }
```

Now, it seems like there's something else about [the @blueprintjs/core `InputGroup` component](https://blueprintjs.com/docs/#core/components/text-inputs.input-group) that's preventing the correct behavior. I'll look into it next.

**Reply by Moshyfawn** [reply]
Alright, so the issue with [the `InputGroup` component](https://blueprintjs.com/docs/#core/components/text-inputs.input-group) is that you want to pass the React Hook Form's `register` `ref` callback into the `inputRef` prop of the `InputGroup` instead of `ref` to forward it to the actual input component.

Also, spreading `{...register}` into a component, you need to forward the `ref` that comes from it into your custom `CustomTextField` component. As your input uses `register`, you opt-in to use the uncontrolled input, so the only way for React Hook Form to `reset` your input's value is by `ref`. So, you need to supply it.

[Here's a quick simplified codesandbox I made with a working example](https://codesandbox.io/s/distracted-archimedes-v6jthw?file=%2Fsrc%2FApp.tsx). Take a look and let me know what you think

**Reply by thooverdev** [reply]
Also of important note, I didn't have my CustomForm wrapped in a < form > tag. Apparently this is pretty important as it won't work without this. TIL how to use forms, thanks @Moshyfawn
